diff -Naur linux-6.15.3.orig/sound/pci/hda/patch_realtek.c linux-6.15.3/sound/pci/hda/patch_realtek.c
--- linux-6.15.3.orig/sound/pci/hda/patch_realtek.c	2025-07-06 12:24:58.380175292 -0400
+++ linux-6.15.3/sound/pci/hda/patch_realtek.c	2025-07-06 12:24:58.869188387 -0400
@@ -11306,6 +11306,7 @@
 	SND_PCI_QUIRK(0x17aa, 0x38fa, "Thinkbook 16P Gen5", ALC287_FIXUP_MG_RTKC_CSAMP_CS35L41_I2C_THINKPAD),
 	SND_PCI_QUIRK(0x17aa, 0x38fd, "ThinkBook plus Gen5 Hybrid", ALC287_FIXUP_TAS2781_I2C),
 	SND_PCI_QUIRK(0x17aa, 0x3902, "Lenovo E50-80", ALC269_FIXUP_DMIC_THINKPAD_ACPI),
+SND_PCI_QUIRK(0x17aa, 0x390b, "Yoga Book 9i GenX", ALC285_FIXUP_SPEAKER2_TO_DAC1),
 	SND_PCI_QUIRK(0x17aa, 0x390d, "Lenovo Yoga Pro 7 14ASP10", ALC287_FIXUP_YOGA9_14IAP7_BASS_SPK_PIN),
 	SND_PCI_QUIRK(0x17aa, 0x3913, "Lenovo 145", ALC236_FIXUP_LENOVO_INV_DMIC),
 	SND_PCI_QUIRK(0x17aa, 0x391f, "Yoga S990-16 pro Quad YC Quad", ALC287_FIXUP_TAS2781_I2C),
