diff -uNr linux-6.15.2.orig/drivers/hid/hid-multitouch.c linux-6.15.2/drivers/hid/hid-multitouch.c
--- linux-6.15.2.orig/drivers/hid/hid-multitouch.c	2025-06-10 07:17:11.000000000 -0400
+++ linux-6.15.2/drivers/hid/hid-multitouch.c	2025-06-18 00:57:50.906678484 -0400
@@ -1761,6 +1761,7 @@
 
 static int mt_probe(struct hid_device *hdev, const struct hid_device_id *id)
 {
+    printk(KERN_INFO "Yoga Book 9i HID driver initialized (patched build)\n");
 	int ret, i;
 	struct mt_device *td;
 	const struct mt_class *mtclass = mt_classes; /* MT_CLS_DEFAULT */
@@ -2146,6 +2147,12 @@
 			   USB_VENDOR_ID_LENOVO,
 			   USB_DEVICE_ID_LENOVO_X12_TAB2) },
 
+
+/* Lenovo Yoga Book 9i Gen X */
+{ .driver_data = MT_CLS_DEFAULT,
+  MT_USB_DEVICE(USB_VENDOR_ID_LENOVO, 0x6161) },
+
+
 	/* Logitech devices */
 	{ .driver_data = MT_CLS_NSMU,
 		HID_DEVICE(BUS_BLUETOOTH, HID_GROUP_MULTITOUCH_WIN_8,
